File: service/src/main/java/org/vsegda/servlet/DataArchiveCronServlet.java
package org.vsegda.servlet;

import org.vsegda.data.DataArchive;
import org.vsegda.data.DataItem;
import org.vsegda.data.DataStream;
import org.vsegda.service.DataItemService;
import org.vsegda.service.DataStreamService;
import org.vsegda.shared.DataStreamMode;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.logging.Logger;

/**
 * @author Roman Elizarov
 */
public class DataArchiveCronServlet extends HttpServlet {
    private static final Logger log = Logger.getLogger(DataArchiveCronServlet.class.getName());

    @Override
    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        log.info("Checking data items for archival needs");
        long startTimeMillis = System.currentTimeMillis();
        for (DataStream stream : DataStreamService.INSTANCE.getDataStreams()) {
            // check for archive or purge (find non-recent items)
            DataItem firstItem = DataItemService.INSTANCE.getFirstDataItem(stream);
            DataItem lastItem = DataItemService.INSTANCE.getLastDataItem(stream);
            log.fine("First item is " + firstItem + "; last item is " + lastItem);
            long threshold =
                    stream.getMode() == DataStreamMode.LAST ? lastItem.getTimeMillis() :
                    System.currentTimeMillis() - DataArchive.RECENT_TIME_INTERVAL - DataArchive.ARCHIVE_INTERVAL;
            if (firstItem != null && firstItem.getTimeMillis() < threshold &&
                    !firstItem.getKey().equals(lastItem.getKey()))
            {
                // need to archive
                DataArchiveTaskServlet.enqueueTask(stream.getStreamId());
            }
        }
        log.info("Done in " + (System.currentTimeMillis() - startTimeMillis) + " ms");
    }
}

File: service/src/main/java/org/vsegda/servlet/DataArchiveCronServlet.kt
package org.vsegda.servlet

import org.vsegda.data.*
import org.vsegda.service.*
import org.vsegda.shared.*
import org.vsegda.util.*
import javax.servlet.http.*

class DataArchiveCronServlet : HttpServlet(), Logged {
    override fun doGet(req: HttpServletRequest, resp: HttpServletResponse) =
        logged("Checking data items for archival needs") {
            for (stream in DataStreamService.dataStreams) {
                // check for archive or purge (find non-recent items)
                val firstItem = DataItemService.getFirstDataItem(stream)
                val lastItem = DataItemService.getLastDataItem(stream)
                log.fine("First item is $firstItem; last item is $lastItem")
                val threshold = if (stream.mode == DataStreamMode.LAST)
                    lastItem.timeMillis
                else
                    System.currentTimeMillis() - RECENT_TIME_INTERVAL - ARCHIVE_INTERVAL
                if (firstItem != null && firstItem.timeMillis < threshold && firstItem.key != lastItem.key) {
                    // need to archive
                    DataArchiveTaskServlet.enqueueTask(stream.streamId)
                }
            }
        }
}
